### ISSUE TITLE
GridSample inputs/outputs not correct

### ISSUE THREAD

**Issue by serwansj** [Tencent/ncnn#4773]
I am trying to export a PyTorch model that contains the `F.grid_sample` operation to NCNN via PNNX. The export works perfectly fine and running it via NCNN in C++ raises no errors, however, the output shape is wrong after the `grid_sample` operation.

The setup is as follows in PyTorch:

```python
class Model(torch.nn.Module):

    def __init__(self):
        super().__init__()

    def forward(self, x):
        in = some_ops(x)    # shape: 8 x 32 x 128 x 128
        grid = some_ops(x)    # shape: 8 x 900 x 4 x 2

        return F.grid_sample(
            in, grid, mode="bilinear", padding_mode="zeros", align_corners=False)
```

The output in PyTorch is of shape `8, 32, 128, 12`.

After i export the model to NCNN via PNNX, I run the following setup in C++ to verify the inputs and outputs:
```cpp
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include "net.h"

int main()
{

    ncnn::Net net;
    net.load_param("models/head_only.ncnn.param");
    net.load_model("models/head_only.ncnn.bin");

    ncnn::Extractor ex = net.create_extractor();

    ncnn::Mat in0(128, 128, 1, 256, sizeof(float));
    ex.input("in0", in0);

    ncnn::Mat in;
    ex.extract("8", in);
    std::cout << in.d << ", " << in.c << ", " << in.h << ", " << in.w << std::endl;

    ncnn::Mat grid;
    ex.extract("6", grid);
    std::cout << grid.d << ", " << grid.c << ", " << grid.h << ", " << grid.w << std::endl;

    ncnn::Mat out;
    ex.extract("out0", out);
    std::cout << out.d << ", " << out.c << ", " << out.h << ", " << out.w << std::endl;

    return 0;
}
```

In NCNN I obtain the following shapes for the inputs and outputs with this setup:
```
in: 32 x 8 x 128 x 128
grid: 900 x 8 x 4 x 2
out: 8 x 8 x 900 x 4
```

To me it seems like during the export, the inputs ended up being formatted wrong since the depth and channel dimensions are swapped in NCNN.

Would appreciate any help on this matter!

**Comment by serwansj** [comment]
Upon further debugging, I concluded that the NCNN `d` and `c` dimensions are swapped in the grid_sample input `in`. However, I am not sure how to fix this issue because this is the result of the PNNX export tool. I can't swap the axes in PyTorch because that would break the code.

**Comment by Yoh-Z** [comment]
emmmm, maybe the output in PyTorch is of shape is [8, 32, 900, 4]？ and if you want to your model work in ncnn, batchsize must be 1 when you convert your model to ncnn by pnnx.

**Comment by serwansj** [comment]
@Yoh-Z i double checked, the output in pytorch is as mentioned in my first post, namely `8, 32, 128, 12` (this is also consistent with Pytorch's documentation about `F.grid_sample`). I made sure batch size is 1 as well.

I managed to fix it manually by changing these two layers in the genrated `.param` file from
```
Permute                  permute_6                1 1 4 6 0=3
Reshape                  reshape_2                1 1 6 70=128 1=128 11=32 2=8
```
to
```
Permute                  permute_6                1 1 4 6 0=13
Reshape                  reshape_2                1 1 6 7 0=128 1=128 11=8 2=32
```

I am still looking for a solution where the export works right out of the box though.

**Comment by serwansj** [comment]
After manually fixing this issue, I ran into https://github.com/Tencent/ncnn/issues/4775. I just realized that both issues are probably caused by mixing the batch dimension with the other dimensions since in my code the inputs to `F.grid_sample` are originally of shape
```
in: 1 x 8 x 32 x 128 x 128
grid: 1 x 8 x 900 x 4 x 2
```
after which the first two dimensions (of which the first one is the batch dimension) are merged to
```
in: 8 x 32 x 128 x 128
grid: 8 x 900 x 4 x 2
```

I am doing this because `F.grid_sample` does not support batched operations. Is there any way to get around this issue? Even if I index
```
in[0]
grid[0]
```
I get a `unbind along batch axis 0 is not supported` error.


So what is the way to go from using batch operations to non-batch operations then?

**Comment by serwansj** [comment]
@Yoh-Z actually, you are right about the output shape in PyTorch, i edited it in my original post. but still, i was using a batch size of 1 during tracing, but the errors persists

**Comment by Yoh-Z** [comment]
> Managed to come up with a minimal example to highlight the problem:
> 
> 1. My PyTorch/Python code is the following:
> 
> ```python
> class Model(torch.nn.Module):
> 
>     def __init__(self):
>         super().__init__()
> 
>     def forward(self, inp, grid):
>         B, N, C, H_in, W_in = inp.shape
>         inp = inp.reshape(B * N, C, H_in, W_in)
>         B, N, H_out, W_out, _ = grid.shape
>         grid = grid.reshape(B * N, H_out, W_out, 2)
>         x = F.grid_sample(
>             inp,
>             grid,
>             mode="bilinear",
>             padding_mode="zeros",
>             align_corners=False)
>         x = x.reshape(B, N, C, H_out, W_out)
>         print(x.shape)
> 
>         return x
> 
> def run():
> 
>     model = Model()
>     model.eval()
>     inp = torch.randn((1, 8, 32, 128, 128))
>     grid = torch.randn((1, 8, 900, 4, 2))
>     mod = torch.jit.trace(model, (inp, grid))
>     mod.save('models/grid_sample.pt')
> ```
> 
> 2. I export to PNNX `./pnnx models/grid_sample.pt inputshape=[1,8,32,128,128],[1,8,900,4,2]` and get the following log:
> 
> ```
> pnnxparam = models/grid_sample.pnnx.param
> pnnxbin = models/grid_sample.pnnx.bin
> pnnxpy = models/grid_sample_pnnx.py
> pnnxonnx = models/grid_sample.pnnx.onnx
> ncnnparam = models/grid_sample.ncnn.param
> ncnnbin = models/grid_sample.ncnn.bin
> ncnnpy = models/grid_sample_ncnn.py
> fp16 = 1
> optlevel = 2
> device = cpu
> inputshape = [1,8,32,128,128]f32,[1,8,900,4,2]f32
> inputshape2 = 
> customop = 
> moduleop = 
> ############# pass_level0
> 
> ----------------
> 
> ############# pass_level1
> ############# pass_level2
> ############# pass_level3
> ############# pass_level4
> ############# pass_level5
> ############# pass_ncnn
> fallback batch axis 233 for operand 0
> fallback batch axis 233 for operand 1
> ```
> 
> 3. Finally, I validate in NCNN:
> 
> ```c++
> #include <opencv2/core/core.hpp>
> #include <opencv2/highgui/highgui.hpp>
> #include "net.h"
> 
> int main()
> {
>     ncnn::Net net;
>     net.load_param("models/grid_sample.ncnn.param");
>     net.load_model("models/grid_sample.ncnn.bin");
> 
>     ncnn::Extractor ex = net.create_extractor();
> 
>     ncnn::Mat in0(128, 128, 8, 32, sizeof(float));
>     ex.input("in0", in0);
>     ncnn::Mat in1(2, 4, 8, 900, sizeof(float));
>     ex.input("in1", in1);
> 
>     ncnn::Mat out;
>     ex.extract("out0", out);
> 
>     std::cout << out.d << ", " << out.c << ", " << out.h << ", " << out.w << std::endl;
>     return 0;
> }
> ```
> 
> and get the output
> 
> ```
> 8, 8, 900, 4
> ```
> 
> , however, the shape should be `8, 32, 900, 4`.
> 
> Hope this example makes this problem easier to debug.

```
inp = torch.randn((1, 8, 32, 128, 128))
grid = torch.randn((1, 8, 900, 4, 2))
```
it isn't the correct shape when you use gridsample in pytorch, and I think it will crash in F.grid_sample.
the shape in 'grid = torch.randn((8, 900, 4, 2))' is (n = 8, c = 900, h = 4, w = 2), and in 'grid = torch.randn((1, 8, 900, 4, 2))' shape is (n = 1, c = 8, d = 900, h = 4, w = 2). when you use 5-d in gridsample, the second Tensor's w mush be 3.
So, you shoule change your model, make n = 1 and the shape is correct, like 'grid = torch.randn((1, 900, 4, 2))'

**Comment by Yoh-Z** [comment]
And if your model must have batchsize = 8, maybe you should use some tirck to make model's batchsize to be 1. 
If your model does not make the batch dimension to operate with other dimensions, it should be easy to do.

**Comment by serwansj** [comment]
my problem is that i have 8 grid sample inputs per batch element, i.e. before the grid sample operation I have an input tensor of shape `B x N x C x H x W` with `B = 1` and `N = 8`. Now I merge the `B` and `N` operations in order to be able to process the inputs with grid_sample, otherwise, it would crash as you pointed out. Here is the reshaping operation that I apply before grid_sample
```
B, N, C, H_in, W_in = inp.shape
inp = inp.reshape(B * N, C, H_in, W_in)
B, N, H_out, W_out, _ = grid.shape
grid = grid.reshape(B * N, H_out, W_out, 2)
```

I assume I need to apply the grid_sample operation in a loop then in NCNN since it does not support batched operations?

**Comment by Yoh-Z** [comment]
https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html
4-d is for (n, c, h, w)
5-d is for (n, c, d, h, w)
what is the (b, n, c, h, w)

**Comment by Yoh-Z** [comment]
if you wanna use 5-d, grid should be (n, d, h, w, 3), but your Tensor is (n, d, h, w, 2) ？

**Comment by Yoh-Z** [comment]
and n is batchsize, so you should get inp = Tensor.randn(1, c, h_in, w_in), grid = Tensor.randn(1, h_out, w_out, 2). pnnx will get right shape and convert it.
like the code in https://github.com/Tencent/ncnn/blob/master/tools/pnnx/tests/test_F_grid_sample.py

**Comment by serwansj** [comment]
I am using the 4-D case of `F.grid_sample`, however, for a single image input (batch size = 1), I have to 8 4-D tensors I need to process by `F.grid_sample`, so effectively for the grid_sample operation I have a batch size 8x the original input batch size.

I tried to do the following now to have a batch size of 1 in `F.grid_sample`, but it does not export correctly (exports as a constant Tensor):

```
out = torch.zeros((1, 8, 32, 900, 4), dtype=grid.dtype, device=grid.device)
for h in range(num_heads):
    out[:, h] = F.grid_sample(inp[:, h], grid[:, h], mode="bilinear", padding_mode="zeros", align_corners=False)
```

Alternatively, I tried
```
out = []
for h in range(num_heads):
    out.append(F.grid_sample(inp[:, h], grid[:, h], mode="bilinear", padding_mode="zeros", align_corners=False))
out = torch.stack(out, dim=1)
```
but also this exports the whole model as a constant tensor.

**Comment by serwansj** [comment]
> And if your model must have batchsize = 8, maybe you should use some tirck to make model's batchsize to be 1.
If your model does not make the batch dimension to operate with other dimensions, it should be easy to do.

What you said here is my main problem I think. My model mixes the batch dimension with the other dimensions, so it's not trivial for me to change it so it has a batch size of 1.

**Comment by serwansj** [comment]
Opened a new issue https://github.com/Tencent/ncnn/issues/4781 for the concat problem.

**Comment by nihui** [comment]
针对onnx模型转换的各种问题，推荐使用最新的pnnx工具转换到ncnn
In view of various problems in onnx model conversion, it is recommended to use the latest pnnx tool to convert your model to ncnn

```shell
pip install pnnx
pnnx model.onnx inputshape=[1,3,224,224]
```

详细参考文档
Detailed reference documentation
https://github.com/pnnx/pnnx
https://github.com/Tencent/ncnn/wiki/use-ncnn-with-pytorch-or-onnx#how-to-use-pnnx

**Comment by nihui** [comment]
The ncnn batch infrastructure feature, which relies on the pnnx tool, is implemented at https://github.com/Tencent/ncnn/commit/127b4146148844ccf67a5bdcb0385e925aa13fbb
